src/main/java/se/liu/ida/hefquin/engine/datastructures/impl/SolutionMappingsHashTable.java

```diff
@@ -1,85 +1,171 @@
 package se.liu.ida.hefquin.engine.datastructures.impl;
 
-import java.util.Iterator;
+import java.util.*;
 
 import org.apache.jena.graph.Node;
 import org.apache.jena.sparql.core.Var;
 
+import org.apache.jena.sparql.engine.binding.Binding;
 import se.liu.ida.hefquin.engine.data.SolutionMapping;
+import se.liu.ida.hefquin.engine.data.impl.SolutionMappingUtils;
 
 public class SolutionMappingsHashTable extends SolutionMappingsIndexBase
 {
+	// Having List<Node> as key type for the hash table is probably
+	// not the best choice in terms of efficiency. However, it will have
+	// to do for the moment.
+	// TODO: can this be made more efficient?
+	protected final Map<List<Node>, List<SolutionMapping>> map = new HashMap<>();
+	protected final Set<Var> joinVariables;
+
+	public SolutionMappingsHashTable(final Set<Var> joinVariables){
+		this.joinVariables = joinVariables;
+	}
+
 	@Override
 	public int size() {
-		// TODO Auto-generated method stub
-		return 0;
+		return map.size();
 	}
 
 	@Override
 	public boolean isEmpty() {
-		// TODO Auto-generated method stub
-		return false;
+		return map.isEmpty();
 	}
 
 	@Override
 	public boolean contains( final Object o ) {
-		// TODO Auto-generated method stub
-		return false;
+		// TODO
+		return map.containsKey(o);
 	}
 
 	@Override
 	public Iterator<SolutionMapping> iterator() {
-		// TODO Auto-generated method stub
-		return null;
+		// The following implementation of this method is inefficient because,
+		// each time it is called, it creates and populates a list of solution
+		// mappings by iterating over the content of this index. However,
+		// for a thread-safe implementation, we may have to live with this.
+		// TODO: is there a more efficient way to implement this method?
+		final List<SolutionMapping> solMap = new ArrayList<>();
+		final Iterator<List<SolutionMapping>> li = map.values().iterator();
+		while(li.hasNext()){
+			solMap.addAll(li.next());
+		}
+		return solMap.iterator();
 	}
 
 	@Override
 	public boolean add( final SolutionMapping e ) {
-		// TODO Auto-generated method stub
-		return false;
+		final List<Node> valKeys = getVarKeys(e);
+
+		if (map.containsKey(valKeys)) {
+			map.get(valKeys).add(e);
+		}
+		else {
+			final List<SolutionMapping> solMapList = new ArrayList<>();
+			solMapList.add(e);
+			map.put(valKeys, solMapList);
+		}
+		return true;
 	}
 
 	@Override
 	public void clear() {
-		// TODO Auto-generated method stub
-		
+		map.clear();
 	}
 
 	@Override
 	public Iterable<SolutionMapping> getJoinPartners( final SolutionMapping sm )
 			throws UnsupportedOperationException
 	{
-		// TODO Auto-generated method stub
-		return null;
+		final List<Node> valKeys = getVarKeys(sm);
+		final List<SolutionMapping> joinPartner = new ArrayList<>();
+		List<SolutionMapping> matchingSolMaps = new ArrayList<>();
+
+		if (!valKeys.isEmpty() && valKeys.size() == joinVariables.size()){
+			matchingSolMaps = map.get(valKeys);
+		}
+		else {
```

Comment:
Instead of `else`, can you do `else if` here, with the condition that `valKeys` is empty.  And then add an `else` branch that throws an `IllegalStateException`.